### PR TITLE
OP: e2e: refactor: extract common logic to `WaitForDelegations`

### PR DESCRIPTION
**Pull request by bap2pecs**
## Summary

I also
- fixed the log util b/c `t.Logf()` will print out the line it's called. so it will always print "log.go Line 9" in the current implementation
- enabled `-race` in our e2e test
- fixed another data race in `GetLastVotedHeight()`

## Test Plan

```
make lint
```

wait for CI

### CODE REVIEW

**Comment by bap2pecs** [comment]
the upstream failed so I will just wait until it's fixed

**Review by lesterli — APPROVED** [review]

**Review by SebastianElvis — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -107,7 +107,7 @@ test-e2e-wasmd: clean-e2e install-babylond install-wasmd
 	@go test -mod=readonly -timeout=25m -v $(PACKAGES_E2E) -count=1 --tags=e2e_wasmd
 
 test-e2e-op: clean-e2e install-babylond
-	@go test -mod=readonly -timeout=25m -v $(PACKAGES_E2E_OP) -count=1 --tags=e2e_op
+	@go test -race -mod=readonly -timeout=25m -v $(PACKAGES_E2E_OP) -count=1 --tags=e2e_op
```
Comment (bap2pecs) [inline Makefile:110]: @SebastianElvis @gusin13 jfyi, after #502, all data races are fixed in our e2e test. so I am enabling it here. You guys can probably do the same thing on your tests.

**Comment by bap2pecs** [comment]
@SebastianElvis I noticed another data race, between `setLastProcessedAndVotedHeight()` and `GetLastVotedHeight()`

the root cause is the lock inside `getStoreFinalityProvider()` is not working as we expected. for example, the below code won't lock the memory for `LastVotedHeight` when it's accessed
```
func (fp *FinalityProviderInstance) GetLastVotedHeight() uint64 {
	return fp.fpState.getStoreFinalityProvider().LastVotedHeight
}

func (fps *fpState) getStoreFinalityProvider() *store.StoredFinalityProvider {
	fps.mu.Lock()
	defer fps.mu.Unlock()
	return fps.fp
}
```

I fixed just this `GetLastVotedHeight()` here to unblock the PR. but we should fix all the other getter functions that has similar pattern in this file


the race:
```
==================
WARNING: DATA RACE
Write at 0x00c007d967c8 by goroutine 128656:
  github.com/babylonchain/finality-provider/finality-provider/service.(*fpState).setLastProcessedAndVotedHeight()
      /Users/zidong/Documents/Projects/babylon-finality-provider/finality-provider/service/fp_store_adapter.go:100 +0x58
  github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).updateStateAfterFinalitySigSubmission()
      /Users/zidong/Documents/Projects/babylon-finality-provider/finality-provider/service/fp_store_adapter.go:162 +0x48
  github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).MustUpdateStateAfterFinalitySigSubmission()
      /Users/zidong/Documents/Projects/babylon-finality-provider/finality-provider/service/fp_store_adapter.go:166 +0x1c
  github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).SubmitBatchFinalitySignatures()
      /Users/zidong/Documents/Projects/babylon-finality-provider/finality-provider/service/fp_instance.go:865 +0x4e8
  github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).FastSync()
      /Users/zidong/Documents/Projects/babylon-finality-provider/finality-provider/service/fastsync.go:83 +0x3ac
  github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).tryFastSync()
      /Users/zidong/Documents/Projects/babylon-finality-provider/finality-provider/service/fp_instance.go:431 +0x1a4
  github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).finalitySigSubmissionLoop()
      /Users/zidong/Documents/Projects/babylon-finality-provider/finality-provider/service/fp_instance.go:248 +0x194
  github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).Start.gowrap1()
      /Users/zidong/Documents/Projects/babylon-finality-provider/finality-provider/service/fp_instance.go:134 +0x34

Previous read at 0x00c007d967c8 by goroutine 705493:
  github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).GetLastVotedHeight()
      /Users/zidong/Documents/Projects/babylon-finality-provider/finality-provider/service/fp_store_adapter.go:127 +0x60
  github.com/babylonchain/finality-provider/itest/opstackl2.(*OpL2ConsumerTestManager).WaitForFpVoteAtHeight.func1()
      /Users/zidong/Documents/Projects/babylon-finality-provider/itest/opstackl2/op_test_manager.go:268 +0x3c
  github.com/stretchr/testify/assert.Eventually.func1()
      /Users/zidong/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1902 +0x38
```

File: itest/opstackl2/op_test_manager.go [hunk 2]
```diff
@@ -68,7 +68,7 @@ func StartOpL2ConsumerManager(t *testing.T) *OpL2ConsumerTestManager {
 	testDir, err := e2eutils.BaseDir("fpe2etest")
 	require.NoError(t, err)
 
-	logger := createLogger(t, zapcore.DebugLevel)
+	logger := createLogger(t, zapcore.ErrorLevel)
```
Comment (bap2pecs) [inline itest/opstackl2/op_test_manager.go:71]: I will revert this back to debug level before merging

File: finality-provider/service/fp_store_adapter.go [hunk 3]
```diff
@@ -124,7 +124,10 @@ func (fp *FinalityProviderInstance) GetStatus() proto.FinalityProviderStatus {
 }
 
 func (fp *FinalityProviderInstance) GetLastVotedHeight() uint64 {
-	return fp.fpState.getStoreFinalityProvider().LastVotedHeight
+	fp.fpState.mu.Lock()
+	lsatVotedHeight := fp.fpState.fp.LastVotedHeight
```
Comment (maurolacy) [inline finality-provider/service/fp_store_adapter.go:128]: Nit:
```suggestion
	lastVotedHeight := fp.fpState.fp.LastVotedHeight
```

File: testutil/log/log.go [hunk 4]
```diff
@@ -1,13 +1,12 @@
...
 import (
-	"testing"
 	"time"
 )
 
-// wraps around testing.T.Logf to add a timestamp prefix
-func Logf(t *testing.T, format string, args ...any) {
+// Add a date prefix to the format string. example usage:
+// t.Logf(log.FormatWithDate("Fp home dir: %s"), fpHomeDir)
+func Prefix(format string) string {
```
Comment (maurolacy) [inline testutil/log/log.go:9]: Can't this format be set in the log at the beginning?

**Review by maurolacy — APPROVED** [review]
LGTM.

File: testutil/log/log.go (hunk 4 above)
Comment (bap2pecs) [reply testutil/log/log.go:9]: yeah unfortunately Go only has a lightweight testing framework and t.Logf doesn't have the feature built-it.

see my conv w runchao on #485

File: itest/opstackl2/op_e2e_test.go [hunk 5]
```diff
@@ -169,18 +169,24 @@ func TestFinalityStuckAndRecover(t *testing.T) {
...
-	time.Sleep(5 * L2BlockTime)
-	// check the finality is stuck
+	t.Logf(log.Prefix("last processed height %d"), lastVotedHeight)
+	// check the finality gets stuck
+	require.Eventually(t, func() bool {
+		latestFinalizedBlock, err := ctm.OpL2ConsumerCtrl.QueryLatestFinalizedBlock()
+		require.NoError(t, err)
+		stuckHeight := latestFinalizedBlock.Height
+		return lastVotedHeight == stuckHeight
+	}, 30*L2BlockTime, e2eutils.EventuallyPollTime)
```
Comment (bap2pecs) [inline itest/opstackl2/op_e2e_test.go:183]: `5 * L2BlockTime` seems still not be enough. I think using an `Eventually` loop is more appropriate

cc @lesterli

**Pull request merged**